### PR TITLE
UIowa Hours

### LINKED ISSUE #4344: As a SNv3 site owner leveraging Exchange to manage resource hours, I would like to be able to display those hours on my website.
# Story

As a SNv3 site owner leveraging https://hours.iowa.uiowa.edu to manage location hours, I would like to be able to display those hours on my website so that end users know when our locations are open/closed and this information can be maintained in one place.

# Proposed solution

Create a block that takes an location identifier (Exchange resource name, ex. RES-foo-200) as an input and passes that to an API and returns the hours for the specified location. The block should have the ability to display open/closed based on the current time compared to the hours of operation.

Related issue: #3933

### PR DESCRIPTION
**Pull request by dnkemper**
Resolves #4344 

## ToDo

- [x] Require `uiowa_hours.settings.group` config be set to place block. (@joewhitsitt)
- [x] Ignore `uiowa_hours.settings`configuration. (@joewhitsitt)
- [x] Make render better (card wrap, fanciness?) - _meeting with Ben to iron this out_
- [x] Status indicator js (@richardbporter)
- [x] ajax layer in drupal announce and other a11y helpers (@joewhitsitt)
- [x] check for "hours" category? (@richardbporter)
- [x] better no results - separate between error and "closed" (@richardbporter)
- [x] default render today for pick a date option
- [x] can we assume one result per resource for a day? the key says otherwise (@richardbporter)
- [x] look through code to make sure everything is named suitably (what is this thing? what will it become?)

# How to test

- `ddev blt sync --site recserv.uiowa.edu`
- Enable hours split and import config.
- Attempt to place block on page, discover that you must select a resource group first before placing a block.
- Go and select a resource group (recserv fusion). Save.
- Place a hours block on a page. Play with the block config options to make sure they work as expected.
- Try the date filter (keyboard, datepicker), after selecting a date, inspect source and find #drupal-live-announce to see that it announces the change to the user.
- ??
- Profit.

### CODE REVIEW

**Comment by joewhitsitt** [comment]
Re-framing this work based on the MauiAPI block. Based on the additions/deletions, this will probably need a sql:drop and sync.

Currently outputs the result from hours.iowa.uiowa.edu in plain text using print_r. Coming back to this in a bit.

**Comment by joewhitsitt** [comment]
Doesn't work yet, but the native date picker looks promising.

Screen Shot 2021-12-08 at 4 55 58 PM

Screen Shot 2021-12-08 at 4 56 59 PM

**Comment by joewhitsitt** [comment]
I think I recall an issue with date pickers and safari...

**Comment by bspeare** [comment]
> I think I recall an issue with date pickers and safari...

https://github.com/uiowa/uiowa/issues/3126

**Edit:** Looks like it is working now.

**Comment by richardbporter** [comment]
Taking a look to familiarize myself with the work done and what is left to do. 👀

**Comment by richardbporter** [comment]
Pushed a pretty big breaking change so this will work for other sites. There is now a config form to set the group. Switched to a select list of resources in block. Probably needs cleaning up.

**Comment by richardbporter** [comment]
>can we assume one result per resource for a day? the key says otherwise

No.

**Comment by richardbporter** [comment]
I have an idea for the status indicators to make it a bit easier. Instead of an option to show or hide a JS-enabled indicator, they would always print along with the hours. 

Screen Shot 2021-12-10 at 7 53 41 AM

This will require refactoring `HoursApi::getHours` to do that as well as handle multiple dates. It would be good to render the date too, also shown in that screenshot.

I can start on this unless someone else has modified that code already.

**Comment by joewhitsitt** [comment]
I’d be open to a JS ninja to remove the jQuery from something I wrote. https://github.com/uiowa/uiowa/pull/4549/commits/a808109fa91bcaf809f7ffe2107c308e7b988397#diff-b3cf0b0ecd6d9e554517802bc0e61b6bf2c99b0724b3938ef1d89ee71ee6c434

**Comment by richardbporter** [comment]
> I’d be open to a JS ninja to remove the jQuery from something I wrote. [a808109#diff-b3cf0b0ecd6d9e554517802bc0e61b6bf2c99b0724b3938ef1d89ee71ee6c434](https://github.com/uiowa/uiowa/commit/a808109fa91bcaf809f7ffe2107c308e7b988397#diff-b3cf0b0ecd6d9e554517802bc0e61b6bf2c99b0724b3938ef1d89ee71ee6c434)

I wouldn't worry about it now. jQuery is everywhere. I think we could remove it all at once when tackling https://github.com/uiowa/uiowa/issues/1977

**Comment by richardbporter** [comment]
I made my changes.

File: docroot/modules/custom/uiowa_hours/src/HoursApi.php [hunk 1]
```diff
@@ -167,28 +169,69 @@ public function getResources($group) {
...
+      foreach ($data as $key => $date) {
+        $render['hours'][$key] = [
+          '#type' => 'container',
+          '#attributes' => [
+            'class' => [
+              'uiowa-hours',
+            ],
+          ],
+          'date' => [
+            '#markup' => $this->t('<h6>@date</h6>', [
```
Comment (richardbporter) [inline docroot/modules/custom/uiowa_hours/src/HoursApi.php]: This should change to another element. Wasn't sure what to use.

Comment (joewhitsitt) [reply docroot/modules/custom/uiowa_hours/src/HoursApi.php]: when in doubt, i've been using `'#type' => 'item',`

Comment (richardbporter) [reply docroot/modules/custom/uiowa_hours/src/HoursApi.php]: Well I mean I think a `h6` is not ideal. Maybe just a `p` with styling?

Comment (joewhitsitt) [reply docroot/modules/custom/uiowa_hours/src/HoursApi.php]: this is now just a string and pulled in as card title

**Comment by joewhitsitt** [comment]
>Warning: strtotime() expects parameter 1 to be string, array given in Drupal\uiowa_hours\HoursApi->getHours() (line 184 of modules/custom/uiowa_hours/src/HoursApi.php).

**Comment by richardbporter** [comment]
>check for "hours" category?

Usage is different between recserv and UHD. Recserv uses categories to distinguish between open/closed while UHD assumes open and uses categories as additional context. However, RecServ said closures aren't currently working and they could get by without them.

**Comment by richardbporter** [comment]
> > Warning: strtotime() expects parameter 1 to be string, array given in Drupal\uiowa_hours\HoursApi->getHours() (line 184 of modules/custom/uiowa_hours/src/HoursApi.php).

Should be fixed.

**Comment by joewhitsitt** [comment]
Started poking at handling multiple resources at once. Not sure it will work out if the expectation is that we also show the "summary" or title of the resource. If the resource is "closed" it is an empty result and so there is nothing but the resource name (e.g. res-rec-crwc) to go off of. Interested in @richardbporter 's thoughts here.

Screen Shot 2021-12-14 at 2 56 34 PM

**Comment by richardbporter** [comment]
> Started poking at handling multiple resources at once. Not sure it will work out if the expectation is that we also show the "summary" or title of the resource. If the resource is "closed" it is an empty result and so there is nothing but the resource name (e.g. res-rec-crwc) to go off of. Interested in @richardbporter 's thoughts here.
> 
> Screen Shot 2021-12-14 at 2 56 34 PM

Good question. I'm not sure to be honest. Have to think about it more.

**Comment by joewhitsitt** [comment]
Holding off on the multiple functionality. It was not originally in-scope, brought up by us, and might have limitations based on what data is available.

Deployed to DEV an update to allow displaying hour summary info alongside the hours. Asked customer to review.

**Comment by joewhitsitt** [comment]
Doing a bit of code clean up and also want to see if there is a better helper function format elsewhere in core/our codebase to base renderResults() on.

**Comment by joewhitsitt** [comment]
Capturing additional future enhancements here: https://github.com/uiowa/uiowa/issues/4570

**Comment by richardbporter** [comment]
I looked this over again. 👍🏻 Might be nice to have someone else review though.

**Comment by bspeare** [comment]
Pulling to test

**Comment by bspeare** [comment]
All the testing criteria checks out.    A few questions:

- The `drupal-live-announce` is a little generic if there are two hours blocks on the page `Returning resource hours information for 2021-12-31`, should the message mention the "Resource" so it says `Returning RES-RECRBTRACK resource hours information for 2021-12-31`?
- The `card_title` is fixed as the default `<h2>`, is it possible to update the `card_title` heading based on the Headline field?

Screen Shot 2021-12-16 at 11 27 16 AM

**Comment by joewhitsitt** [comment]
>The card_title is fixed as the default `<h2>`, is it possible to update the card_title heading based on the Headline field?

Ah yes, I will see about wiring that up.

As for the announce information, not sure if the resource name will be very helpful to a casual visitor. I will see about incorporating the block title if provided.

Additionally, it isn't in the announce section but the ajax wrapper has an aria live setup and should read off the changes/hours information. If they include the summary, that should also be announced. You can test this with VoiceOver or watch me attempt it in demo.

**Comment by joewhitsitt** [comment]
Made some changes to get the child headline to work. Re-organized some things in the process so I will be testing live at demo!

**Comment by bspeare** [comment]
@joewhitsitt The child heading sizes don't seem to be changing for me still.   Maybe someone else can check that out to see if it's just my local environment.

**Comment by joewhitsitt** [comment]
Thanks for checking. I _think_ it is working. At least for me.

Screen Shot 2021-12-17 at 8 35 17 AM

**Comment by e-marie-w** [comment]
I'm also not seeing the child headlines adjust when the headline size is changed. I'm getting some 'undefined index' errors related to the headlines too. Not sure if the errors are part of this or just my local environment.

**Comment by joewhitsitt** [comment]
@e-marie-w when you get a chance, can you pull an `cr` to see if those issues have been resolved?

**Comment by e-marie-w** [comment]
@joewhitsitt I still have the errors, but the child headline sizing works fine from the start, including the closed results that weren't adjusting correctly before. Without a headline the child content headline sizing still starts at heading 3, was that what we wanted?

**Comment by joewhitsitt** [comment]
heading three default is correct. the block assumes it should be nested under a headline elsewhere if not using the headline in the block. also "child heading" should show initially for this block similar to articles

h1 (page title), some identifying content (h2), hours date (h3)

**Comment by e-marie-w** [comment]
Okay, awesome! And the hours block does work fine on a new page, no errors at all. It was just the existing page with existing blocks causing issues apparently.

**Review by e-marie-w — APPROVED** [review]

**Pull request merged**